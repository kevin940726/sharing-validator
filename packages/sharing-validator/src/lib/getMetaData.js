import fetch from "node-fetch";
import { Parser } from "htmlparser2";
import USER_AGENTS from "./userAgents";

async function getMetadata(
  url,
  { userAgent = USER_AGENTS.general, headers = {} } = {}
) {
  const response = await fetch(url, {
    headers: {
      "User-Agent": userAgent,
      ...headers
    }
  });

  let isTitle = false;
  let title = "";

  return new Promise(resolve => {
    const meta = [];

    const htmlParser = new Parser(
      {
        onopentag(name, attrs) {
          if (name === "meta" && (attrs.property || attrs.name)) {
            meta.push([attrs.property || attrs.name, attrs.content]);
          } else if (name === "title") {
            isTitle = true;
            title = "";
          }
        },
        onText(text) {
          if (isTitle) {
            title += text;
          }
        },
        onclosetag(name) {
          if (name === "head") {
            htmlParser.end();
          } else if (name === "title" && isTitle) {
            if (title) {
              meta.push(["title", title]);
            }
            isTitle = false;
          }
        },
        onend() {
          resolve(meta);
        }
      },
      { decodeEntities: false }
    );

    response.body.pipe(htmlParser);
  });
}

export default getMetadata;
