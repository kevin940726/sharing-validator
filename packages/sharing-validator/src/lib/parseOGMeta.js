function parseMetaData(meta) {
  const data = {};

  for (let [key, value] of meta) {
    if (!key || key === "undefined") {
      // Skip meta which doesn't have a `property` or `name` attribute
      continue;
    }

    const properties = key.split(":");

    let target = data;
    for (let property of properties.slice(0, -1)) {
      if (typeof target[property] === "string") {
        target[property] = {
          url: target[property]
        };
      } else {
        target[property] = target[property] || {};
      }

      target = target[property];

      if (Array.isArray(target)) {
        target = target.slice(-1)[0];
      }
    }

    const lastKey = properties.slice(-1)[0];
    const originalValue = Array.isArray(target[lastKey])
      ? target[lastKey][0]
      : target[lastKey];

    if (originalValue) {
      if (typeof originalValue === "string") {
        target[lastKey] = [].concat(target[lastKey], value);
      } else {
        target[lastKey] = [].concat(target[lastKey], { url: value });
      }
    } else {
      target[lastKey] = value;
    }
  }

  return data;
}

export default parseMetaData;
