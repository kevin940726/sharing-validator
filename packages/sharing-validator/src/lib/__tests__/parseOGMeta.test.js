import parseOGMeta from "../parseOGMeta";
import getMetaData from "../getMetaData";

test("baseline", () => {
  const meta = [
    ["undefined", "ie=edge"],
    ["viewport", "width=device-width, initial-scale=1, shrink-to-fit=no"],
    ["generator", "Gatsby 2.15.11"],
    ["description", "I&#x27;m Kai Hao, a front-end developer in Taiwan."],
    ["og:url", "https://kaihao.dev/"],
    ["og:title", "Kai Hao"],
    ["og:description", "I&#x27;m Kai Hao, a front-end developer in Taiwan."],
    [
      "og:image",
      "https://kaihao.dev/static/15baf5b2b73048942a827496d3805470/70e2c/default-meta-image.png"
    ],
    ["og:image:width", "1200"],
    ["og:image:height", "626"],
    ["og:type", "website"],
    ["twitter:card", "summary_large_image"],
    ["twitter:creator", "@kevin940726"],
    ["twitter:title", "Kai Hao"],
    [
      "twitter:description",
      "I&#x27;m Kai Hao, a front-end developer in Taiwan."
    ],
    [
      "twitter:image",
      "https://kaihao.dev/static/15baf5b2b73048942a827496d3805470/70e2c/default-meta-image.png"
    ],
    ["theme-color", "#f2994a"]
  ];

  expect(parseOGMeta(meta)).toMatchSnapshot();
});

// https://ogp.me/#array
test("array", () => {
  const meta = [
    ["og:image:width", "300"],
    ["og:image", "http://example.com/rock.jpg"],
    ["og:image:height", "300"],
    ["og:image", "http://example.com/rock2.jpg"],
    ["og:image", "http://example.com/rock3.jpg"],
    ["og:image:height", "1000"]
  ];

  expect(parseOGMeta(meta)).toEqual({
    og: {
      image: [
        { url: "http://example.com/rock.jpg", width: "300", height: "300" },
        { url: "http://example.com/rock2.jpg" },
        { url: "http://example.com/rock3.jpg", height: "1000" }
      ]
    }
  });
});
