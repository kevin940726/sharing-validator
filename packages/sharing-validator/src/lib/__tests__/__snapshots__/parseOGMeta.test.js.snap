// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseline 1`] = `
Object {
  "description": "I&#x27;m Kai Hao, a front-end developer in Taiwan.",
  "generator": "Gatsby 2.15.11",
  "og": Object {
    "description": "I&#x27;m Kai Hao, a front-end developer in Taiwan.",
    "image": Object {
      "height": "626",
      "url": "https://kaihao.dev/static/15baf5b2b73048942a827496d3805470/70e2c/default-meta-image.png",
      "width": "1200",
    },
    "title": "Kai Hao",
    "type": "website",
    "url": "https://kaihao.dev/",
  },
  "theme-color": "#f2994a",
  "twitter": Object {
    "card": "summary_large_image",
    "creator": "@kevin940726",
    "description": "I&#x27;m Kai Hao, a front-end developer in Taiwan.",
    "image": "https://kaihao.dev/static/15baf5b2b73048942a827496d3805470/70e2c/default-meta-image.png",
    "title": "Kai Hao",
  },
  "viewport": "width=device-width, initial-scale=1, shrink-to-fit=no",
}
`;

exports[`spotify 1`] = `
Object {
  "al": Object {
    "android": Object {
      "app_name": "Spotify",
      "package": "com.spotify.music",
      "url": "spotify://track/4u7EnebtmKWzUH433cf5Qv",
    },
    "ios": Object {
      "app_name": "Spotify",
      "app_store_id": "324684580",
      "url": "spotify://track/4u7EnebtmKWzUH433cf5Qv",
    },
  },
  "fb": Object {
    "app_id": "174829003346",
  },
  "google": "notranslate",
  "music": Object {
    "album": Object {
      "track": "11",
      "url": "https://open.spotify.com/album/1GbtB4zTqAsyfZEsm1RZfx",
    },
    "duration": "354",
    "musician": "https://open.spotify.com/artist/1dfeR4HaWDbWqFHLkxsg1d",
    "preview_url": Object {
      "secure_url": "https://p.scdn.co/mp3-preview/d1ff0ba5c5538ca2c50b808aab2278253c98b038?cid=162b7dc01f3a4a2ca32ed3cec83d1e02",
      "type": "audio/mpeg",
      "url": "http://p.scdn.co/mp3-preview/d1ff0ba5c5538ca2c50b808aab2278253c98b038?cid=162b7dc01f3a4a2ca32ed3cec83d1e02",
    },
    "release_date": "1975-11-21",
  },
  "og": Object {
    "audio": Object {
      "type": "audio/vnd.facebook.bridge",
      "url": "https://p.scdn.co/mp3-preview/d1ff0ba5c5538ca2c50b808aab2278253c98b038?cid=162b7dc01f3a4a2ca32ed3cec83d1e02",
    },
    "description": "Bohemian Rhapsody - Remastered 2011, a song by Queen on Spotify",
    "image": "https://i.scdn.co/image/356e29a0f112a4835f7b22f5e586a1907eae2031",
    "restrictions": Object {
      "country": Object {
        "allowed": Array [
          "AD",
          "AE",
          "AR",
          "AT",
          "AU",
          "BE",
          "BG",
          "BH",
          "BO",
          "BR",
          "CA",
          "CH",
          "CL",
          "CO",
          "CR",
          "CY",
          "CZ",
          "DE",
          "DK",
          "DO",
          "DZ",
          "EC",
          "EE",
          "EG",
          "ES",
          "FI",
          "FR",
          "GB",
          "GR",
          "GT",
          "HK",
          "HN",
          "HU",
          "ID",
          "IE",
          "IL",
          "IN",
          "IS",
          "IT",
          "JO",
          "JP",
          "KW",
          "LB",
          "LI",
          "LT",
          "LU",
          "LV",
          "MA",
          "MC",
          "MT",
          "MX",
          "MY",
          "NI",
          "NL",
          "NO",
          "NZ",
          "OM",
          "PA",
          "PE",
          "PH",
          "PL",
          "PS",
          "PT",
          "PY",
          "QA",
          "RO",
          "SA",
          "SE",
          "SG",
          "SK",
          "SV",
          "TH",
          "TN",
          "TR",
          "TW",
          "US",
          "UY",
          "VN",
          "ZA",
        ],
      },
    },
    "site_name": "Spotify",
    "title": "Bohemian Rhapsody - Remastered 2011",
    "type": "music.song",
    "url": "https://open.spotify.com/track/4u7EnebtmKWzUH433cf5Qv",
  },
  "twitter": Object {
    "app": Object {
      "id": Object {
        "googleplay": "com.spotify.music",
        "iphone": "324684580",
      },
    },
    "audio": Object {
      "artist_name": "Queen",
      "partner": "Spotify",
      "source": "https://open.spotify.com/twitter/vmap/track/4u7EnebtmKWzUH433cf5Qv?utm_campaign=twitter-player&amp;utm_source=open&amp;utm_medium=twitter",
    },
    "card": "audio",
    "description": "Bohemian Rhapsody - Remastered 2011, a song by Queen on Spotify",
    "image": "https://i.scdn.co/image/356e29a0f112a4835f7b22f5e586a1907eae2031",
    "player": Array [
      Object {
        "height": "584",
        "width": "504",
      },
      Object {
        "url": "https://open.spotify.com/embed/track/4u7EnebtmKWzUH433cf5Qv?utm_campaign=twitter-player&amp;utm_source=open&amp;utm_medium=twitter",
      },
    ],
    "site": "@spotify",
    "title": "Bohemian Rhapsody - Remastered 2011",
  },
}
`;
